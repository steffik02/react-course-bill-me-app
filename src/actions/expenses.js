import uuid from 'uuid';
import database from '../firebase/firebase';

        //use push
        //attach then callback
        //dispatch action
        //redirect

// add expense
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            description = '', 
            note = '', 
            amount = 0, 
            createdAt = 0 
        } = expenseData;
        const expense = { description, note, amount, createdAt }
        return database.ref(`users/${uid}/expenses`).push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        })
    }
}

// remove expense
export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

//create new action with same call signature as removeExpense
//test startRemoveExpense with 'should remove expenses from firebase'
    // to test this you can just try to fetch it, call val on the snapshot, and if no data null will be return value
//use startRemoveExpense in EditExpensePage instead of removeExpense
//Adjust EditExpensePage tests
export const startRemoveExpense = ({ id } = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).remove().then(() => {
            dispatch(removeExpense({id}));
        });
    };
};

// edit expense
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

export const startEditExpense = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).update(updates).then(() => {
            dispatch(editExpense(id, updates));
        });
    };
};

// SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

// fetch all expense data from the db once
// parse the data into an array (see firebase.js file)
// dispatch SET_EXPENSES to change the data in the store
export const startSetExpenses = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses`).once('value').then((snapshot) => {
            const expenses = [];
            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
            dispatch(setExpenses(expenses));
        });
    };
};
