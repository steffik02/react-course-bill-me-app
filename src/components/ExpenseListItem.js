import React from 'react';
import { Link } from "react-router-dom";
import moment from 'moment';
import numeral from 'numeral';

//connect component to access dispatch
//import correct delete functin
//wire up onClick

export default ({id, description, amount, createdAt}) => (
    <div>
        <Link to={'/edit/' + id}>
            <h3>Description: {description}</h3>
        </Link>
        <p>Amount: {numeral(amount / 100).format('$0,0.00')}</p>
        <p>Date Due: {moment(createdAt).format('MMMM D, YYYY')}</p>
    </div>
);