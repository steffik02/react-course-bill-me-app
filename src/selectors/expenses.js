import moment from 'moment';

// Get visible expenses
export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true; 
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const lcdesc = expense.description.toLowerCase();
        const lctext = text.toLowerCase();
        const textMatch = lcdesc.includes(lctext);
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1
        } else if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1
        };
    });
};