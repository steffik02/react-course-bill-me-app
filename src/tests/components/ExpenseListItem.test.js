import React from 'react';
import ExpenseListItem from '../../components/ExpenseListItem';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';

test('Should render ExpenseListItem with one expense', () => {
    const wrapper = shallow(<ExpenseListItem {...expenses[1]} />);
    expect(wrapper).toMatchSnapshot();
});