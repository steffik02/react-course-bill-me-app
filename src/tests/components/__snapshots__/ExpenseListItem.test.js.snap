// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpenseListItem with one expense 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/2"
  >
    <h3>
      Description: 
      Rent
    </h3>
  </Link>
  <p>
    Amount: 
    $1,095.00
  </p>
  <p>
    Date Due: 
    December 27, 1969
  </p>
</div>
`;
